metaspace2020/metaspace
# Feat/imzml browser

### Description

A new feature which makes possible to browse the dataset's imzml file, by making possible to visualize the ion Image and the selected pixel spectrum chart.
#582 

### How to test it

As the imzml browser backend routes are a POC, in order to test the front-end, you need to run the backend as a standalone server. To test it, follow these steps:

1 - Run the metaspace project
2 - Go to the the folder metaspace/metaspace/browser and register the sm dev package:  python3 setup.py develop
3 - Go to the folder /metaspace/metaspace/browser/sm/browser and run the server: uvicorn api:app --reload 
4 - Select the inputPath of the dataset you want to check the functionality, add its value to the attribute s3_path and pre-process it (via postman):
     route: http://127.0.0.1:8000/preprocess
     body: {
             "s3_path": "s3://sm-engine-dev/ddd7f0e5-ff74-4f61-96d4-60f40dfb4c3c",
     }
     * This step is needed, because I still didn't create a back-end route to check if the temp files generated already exists (i didnt do because I dont know how this will be in the production env).
5 - After pre-process the dataset, you can access the imzml browser page on the front-end: http://localhost:8999/dataset/<dataset_id>/browser    

#### Changes

##### Webapp

###### DatasetBrowserPage
- [x] Added spectrum chart filters
> - [x] Filter by database
> - [x] Filter by only annotated by FDR
- [x] Added image filters
> - [x] Filter by m/z
> - [x] Filter by formula
- [x] Querying ion image by s3_path, mz and ppm
- [x] Querying spectrum chart by selected x,y pixel
To do:
- [ ] Check if imzml temp files exists before doing the search query
- [ ] Add TIC normalization
- [ ] Handle back-end errors
- [ ] Normalize ion image with max and min mz returned from backend

###### DatasetBrowserSpectrumChart
- [x] Showing instructions to visualize the chart when empty
- [x] Showing spectrum chart
- [x] Added spectrum chart image download feature
- [x] Added spectrum y axis zoom by slider
- [x] Added spectrum x axis zoom by mouse scroll
- [x] Added chart zoom full reset by double clicking the chart
- [x] Added chart zoom full reset by clicking the restore button
- [x] Added chart zoom last step reset by clicking the reset button
- [x] Added chart zoom with brushing tool by clicking the zoom button

###### Dataset api
- [x] Returning input_path in datasetDetail fragment


###### Ion image viewer
- [x] Added keep pixel selected logic and callback
To do:
- [ ] Move selected pixel by pressing keyboard arrows


###### Simple ion image viewer
- [x] Created simple wrapper to ion image with embed intensity slider handler

##### Browser

###### Api
- [x] Enabled CORS for testing on localhost
- [x] Added search_pixel route to get spectrum data by passing pixel coordinates

###### Main
- [x] Added search_pixel by spectrum function

###### utils
- [x] Enabled minio usage


#### Tests
 
##### Front end
 
 
###### Browsers and resolutions



#### Considerations
* A warning from echarts is displayed on the console ("import echarts from 'echarts/lib/echarts'" is not supported anymore. Use "import * as echarts from 'echarts/lib/echarts'" instead;) as the vue-echarts wrapper is importing it on the old way. But according to the code this warning is not shown on production mode. I can downgrade echarts version so this warning is not shown, but when downgraded the label overlapping function does not work.
* Apparently, by using onResult feature from useQuery, even when setting the localVue, jest does not seem to be able to parse the file, so thats why i removed the DatasetBrowserPage tests for now.
